component/libfuse/libfuse_handler.go

```diff
@@ -0,0 +1,1178 @@
+		if err == syscall.ENOENT {
+			return -C.ENOENT
+		} else if err == syscall.EACCES {
+			return -C.EACCES
+		} else {
+			return -C.EIO
+		}
+	}
+
+	// Populate stat
+	fuseFS.fillStat(attr, stbuf)
+	return 0
+}
+
+// Directory Operations
+
+// libfuse_mkdir creates a directory
+//
+//export libfuse_mkdir
+func libfuse_mkdir(path *C.char, mode C.mode_t) C.int {
+	name := trimFusePath(path)
+	name = common.NormalizeObjectName(name)
+	log.Trace("Libfuse::libfuse_mkdir : %s", name)
+
+	err := fuseFS.NextComponent().CreateDir(internal.CreateDirOptions{Name: name, Mode: fs.FileMode(uint32(mode) & 0xffffffff)})
+	if err != nil {
+		log.Err("Libfuse::libfuse_mkdir : Failed to create %s [%s]", name, err.Error())
+		if os.IsPermission(err) {
+			return -C.EACCES
+		} else if os.IsExist(err) {
+			return -C.EEXIST
+		} else {
+			return -C.EIO
+		}
+	}
+
+	libfuseStatsCollector.PushEvents(createDir, name, map[string]interface{}{md: fs.FileMode(uint32(mode) & 0xffffffff)})
+	libfuseStatsCollector.UpdateStats(stats_manager.Increment, createDir, (int64)(1))
+
+	return 0
+}
+
+// libfuse_opendir opens handle to given directory
+//
+//export libfuse_opendir
+func libfuse_opendir(path *C.char, fi *C.fuse_file_info_t) C.int {
+	name := trimFusePath(path)
+	name = common.NormalizeObjectName(name)
+	if name != "" {
+		name = name + "/"
+	}
+
+	log.Trace("Libfuse::libfuse_opendir : %s", name)
+
+	handle := handlemap.NewHandle(name)
+
+	// For each handle created using opendir we create
+	// this structure here to hold current block of children to serve readdir
+	handle.SetValue("cache", &dirChildCache{
+		sIndex:   0,
+		eIndex:   0,
+		token:    "",
+		length:   0,
+		children: make([]*internal.ObjAttr, 0),
+	})
+
+	handlemap.Add(handle)
+	fi.fh = C.ulong(uintptr(unsafe.Pointer(handle)))
+
+	return 0
+}
+
+// libfuse_releasedir opens handle to given directory
+//
+//export libfuse_releasedir
+func libfuse_releasedir(path *C.char, fi *C.fuse_file_info_t) C.int {
+	handle := (*handlemap.Handle)(unsafe.Pointer(uintptr(fi.fh)))
+
+	log.Trace("Libfuse::libfuse_releasedir : %s, handle: %d", handle.Path, handle.ID)
+
+	handle.Cleanup()
+	handlemap.Delete(handle.ID)
+	return 0
+}
+
+// libfuse_readdir reads a directory
+//
+//export libfuse_readdir
+func libfuse_readdir(_ *C.char, buf unsafe.Pointer, filler C.fuse_fill_dir_t, off C.off_t, fi *C.fuse_file_info_t, flag C.fuse_readdir_flags_t) C.int {
+	handle := (*handlemap.Handle)(unsafe.Pointer(uintptr(fi.fh)))
+
+	handle.RLock()
+	val, found := handle.GetValue("cache")
+	handle.RUnlock()
+
+	if !found {
+		return C.int(C_EIO)
+	}
+
+	off_64 := uint64(off)
+	cacheInfo := val.(*dirChildCache)
+	if off_64 == 0 ||
+		(off_64 >= cacheInfo.eIndex && cacheInfo.token != "") {
+		attrs, token, err := fuseFS.NextComponent().StreamDir(internal.StreamDirOptions{
+			Name:   handle.Path,
+			Offset: off_64,
+			Token:  cacheInfo.token,
+			Count:  common.MaxDirListCount,
+		})
```

Comment:
Good point. Yes probably good as a backlog. I took a look at the changes for readdir and it would be a bit more involved, particularly since fuse3 is using cgo so it is not quite as simple to fill in the buffer for the list of items.